amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -454,15 +454,17 @@ Set<StreamIdentifier> checkAndSyncStreamShardsAndLeases()
             final MetricsScope metricsScope = MetricsUtil.createMetricsWithOperation(metricsFactory, MULTI_STREAM_TRACKER);
 
             try {
+                final Map<StreamIdentifier, StreamConfig> newStreamConfigMap = streamTracker.streamConfigList()
+                        .stream().collect(Collectors.toMap(StreamConfig::streamIdentifier, Function.identity()));
                 // This is done to ensure that we clean up the stale streams lingering in the lease table.
-                if (!leasesSyncedOnAppInit && isMultiStreamMode) {
-                    final List<MultiStreamLease> leases = fetchMultiStreamLeases();
-                    syncStreamsFromLeaseTableOnAppInit(leases);
-                    leasesSyncedOnAppInit = true;
+                if (shouldSyncLeases && isMultiStreamMode) {
+                    // Skip updating the stream map due to no new stream since last sync
+                    if (newStreamConfigMap.keySet().stream().anyMatch(s -> !currentStreamConfigMap.containsKey(s))) {
+                        syncStreamsFromLeaseTable(fetchMultiStreamLeases());
+                    }
+                    shouldSyncLeases = false;
```

Comment:
this would mean we will be doing the perfoming leaseSyncs on every host now which would lead to increase in the Lease table IOPs significantly, specifically on deployment.

I think we should consider evaluating the solution where on new leader itself we evaluate if shardSync is needed (rather than doing shardSync blindly) if not needed then simply update the in-memory state else perform shardSync(as this is requirement anyways). 

Check the logic in PeriodicShardSyncManager which is optimized logic and does not lead to many scan operations but also perform shardSync only when needed. Today in Scheduler we are assuming if leader is not info of stream then shardSync is needed which is not correct and can be optimized IMHO. Thoughts?

Reply:
Letting `PeriodicShardSyncManager` to do all the shard sync in one place is also the ideal solution I have in my mind as pointed out in the overview that PeriodicShardSyncManager already has all the functionality but here are my concerns around going to that approach directly. 

1. Incremental change - this change has smaller blast radius and provides IOPs improvements already in most of the cases.
2. Backward compatibility - the logic we have now is that leader learns new streams and enqueues directly which might be faster than letting the periodic shard sync manager to start doing shard sync. KCL users might set really high wait time between each periodic shard sync runs and this would cause delay in learning and crating lease for the new streams.

Reply:
As per offline discussion, looks good and this should not have the IOPs increase with deployment as the change is on the leader side only and not on each host. 

Logic looks good, Lets test this change also properly to ensure we do not miss or add any edge condition.

Reply:
Have performed and documented two tests.

T1: Once a worker is being re-elected for leader it can do DDB lookup to update its in memory map if detecting  new stream.

T2: During the lifetime of a leader the DDB lookup is executed at most once.